cmake/IBAMRTesting.cmake

```diff
@@ -0,0 +1,285 @@
+        set("restart" ${_x_value})
+      else (_x_key STREQUAL "restart")
+        message(WARNING "Encountered unrecognized override ${_x_key}")
+      endif ()
+      message(STATUS "${_x_key} ${_x_value}")
+    endif ()
+  endforeach ()
+
+  set(n_mpirun ${n_mpirun} PARENT_SCOPE)
+  set(n_restart ${n_restart} PARENT_SCOPE)
+  set(expect_error ${expect_error} PARENT_SCOPE)
+endfunction()
+
+#[[!md
+
+ibamr_add_testss(
+  base_name
+  DIRECTORY
+    TestDirectory
+)
+
+]]#
+function (ibamr_add_tests name)
+  cmake_parse_arguments(test
+    ""
+    "DIRECTORY"
+    "LIBRARIES;SOURCES" ${ARGN})
+
+  if (NOT test_DIRECTORY)
+    message(FATAL_ERROR "Missing DIRECTORY for test")
+  endif ()
+
+  if (NOT IBAMR_ENABLE_TESTING)
+    return ()
+  endif ()
+
+  #message(CHECK_START "Adding tests: ${test_DIRECTORY}/${name}")
+  #list(APPEND CMAKE_MESSAGE_INDENT "  ")
+
+  # Add test source if it exists (may not because of _2d/_3d suffix)
+  if (EXISTS "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+    list(APPEND test_SOURCES "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+  endif ()
+
+  # If there aren't any sources, the this is not going to work...
+  if (NOT test_SOURCES)
+    message(FATAL_ERROR "No sources detected for test: ${test_DIRECTORY}/${name}")
+  endif ()
+
+  # Find all of the test inputs
+  file(GLOB input_files "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.*.input")
+  if (NOT input_files)
+    #list(POP_BACK CMAKE_MESSAGE_INDENT)
+    #message(CHECK_FAIL "no test inputs detected, skipping...")
+    return ()
+  endif ()
+
+  ###############################################################################################
+  # Create build targets
+
+  # Add a top level test build target
+  if (NOT TARGET "ibamr-tests")
+    add_custom_target(ibamr-tests ALL)
+  endif ()
+
+  # Add a direcotry build target and make the top level depend on it
+  if (NOT TARGET "ibamr-tests-${test_DIRECTORY}")
+    add_custom_target("ibamr-tests-${test_DIRECTORY}")
+    add_dependencies(ibamr-tests "ibamr-tests-${test_DIRECTORY}")
+  endif ()
+
+  # Create the test executable
+  set(_target "ibamr-test-${test_DIRECTORY}-${name}")
+  add_executable(${_target} EXCLUDE_FROM_ALL "${test_SOURCES}")
+  set_target_properties(${_target}
+    PROPERTIES
+    RUNTIME_OUTPUT_DIRECTORY
+      "${CMAKE_BINARY_DIR}/tests/${_dir}"
+    OUTPUT_NAME
+      ${name}
+    )
+  target_link_libraries(${_target} PRIVATE ${test_LIBRARIES})
+  target_compile_definitions(${_target}
+    PRIVATE
+      SOURCE_DIR="${IBAMR_TEST_DIR}/${test_DIRECTORY}/")
+  # Connect the target to the directory target
+  add_dependencies("ibamr-tests-${test_DIRECTORY}" ${_target})
+
+  # Add the tests for this target
+  foreach (input ${input_files})
+    get_filename_component(input_stripped ${input} NAME_WLE)
+    get_filename_component(input_directory ${input} DIRECTORY)
+    set(check_output_file "${input_directory}/${input_stripped}.output")
+
+    _ibamr_parse_test(${input_stripped})
```

Comment:
So, that is a good habit, and that is what the intention is. This function should only be called in this file, but CMake doesn't have a `static` equivalent that I know of. My solution, add an `_` in front of the function name that says it shouldn't be called elsewhere :)

Reply:
Got it. It's not that important (we can always homogenize the style later) but in that case the `_ibamr` is implied.

The other style habit I have is to use all capital letters for macros and functions (again, this is inherited from other projects I have worked on) so that all function calls look the same.

Reply:
If we don't export the macros then I suppose we don't need to preface them with `IBAMR_` - I don't think I was fully consistent in this when I wrote the rest of the build system.

Reply:
This style is taken from Kitware projects. I wasn't sure what you would want to do since there weren't really any other CMake functions like this. I will go back and make all of the public CMake test API functions upper case.